core/src/main/java/com/google/errorprone/bugpatterns/DefaultLocale.java

```diff
@@ -0,0 +1,698 @@
+            state.getSourceForNode(ASTHelpers.getReceiver(tree)),
+            SuggestedFixes.qualifyStaticImport("java.text.DateFormat.SHORT", fix, state),
+            localeFix.replacement(fix, state)));
+    return fix.build();
+  }
+
+  private void dateFormatFixes(
+      Description.Builder description,
+      MethodInvocationTree tree,
+      VisitorState state,
+      int nonLocaleArgs) {
+    description.addFix(dateFormatFix(tree, state, nonLocaleArgs, LocaleFix.ROOT_FIX));
+    description.addFix(dateFormatFix(tree, state, nonLocaleArgs, LocaleFix.DEFAULT_LOCALE_FIX));
+    description.addFix(
+        dateFormatFix(tree, state, nonLocaleArgs, LocaleFix.DEFAULT_FORMAT_LOCALE_FIX));
+  }
+
+  private Fix dateFormatFix(
+      MethodInvocationTree tree, VisitorState state, int nonLocaleArgs, LocaleFix localeFix) {
+    var fix = SuggestedFix.builder().setShortDescription(localeFix.title());
+    if (tree.getArguments().isEmpty()) {
+      var defaultConst =
+          SuggestedFixes.qualifyStaticImport("java.text.DateFormat.DEFAULT", fix, state);
+      fix.replace(
+          state.getEndPosition(tree.getMethodSelect()),
+          state.getEndPosition(tree),
+          String.format(
+              "(%s, %s)",
+              Stream.generate(() -> defaultConst)
+                  .limit(nonLocaleArgs)
+                  .collect(Collectors.joining(", ")),
+              localeFix.replacement(fix, state)));
+    } else {
+      fix.postfixWith(
+          Iterables.getLast(tree.getArguments()), ", " + localeFix.replacement(fix, state));
+    }
+    return fix.build();
+  }
+
+  private Description handleMessageFormatFormat(MethodInvocationTree tree, VisitorState state) {
+    var pattern = tree.getArguments().get(0);
+    var arguments = tree.getArguments().stream().skip(1).collect(toImmutableList());
+    if (!shouldRefactorStringFormat(pattern, arguments, state)) {
+      return NO_MATCH;
+    }
+    var description = buildDescription(tree);
+    description.addFix(messageFormatFormatFix(tree, pattern, arguments, state, LocaleFix.ROOT_FIX));
+    description.addFix(
+        messageFormatFormatFix(tree, pattern, arguments, state, LocaleFix.DEFAULT_LOCALE_FIX));
+    description.addFix(
+        messageFormatFormatFix(
+            tree, pattern, arguments, state, LocaleFix.DEFAULT_FORMAT_LOCALE_FIX));
+    return description.build();
+  }
+
+  /**
+   * Should only refactor String.format() and similar methods' invocations where specifiers aren't
+   * locale-dependent. For %s this includes checking for non-Formattable arguments. Format strings
+   * (first argument) as variables or constants are excluded from refactoring.
+   */
+  private boolean shouldRefactorStringFormat(
+      ExpressionTree pattern, List<? extends ExpressionTree> arguments, VisitorState state) {
+    if (!(pattern instanceof LiteralTree)) return true;
+    if (!onlyContainsSpecifiersInAllowList((String) ((LiteralTree) pattern).getValue()))
```

Comment:
This is only matching on string literals, not on other compile-time constant expressions like static final constant fields.

The following will handle compile-time constant expressions:

```
String patternValue = ASTHelpers.constValue(pattern, String.class)
if (patternValue != null) { ... }
```

Reply:
A related question is the behaviour if the format string isn't available at compile time. The current behaviour rejects that, I mentioned the possibility of allowing it in the bug. It comes down to a tradeoff between soundness and false negatives.

Reply:
When changing to `ASTHelpers.constValue`, I actually made it to allow non-constant patterns. I added a test, and a TODO to possibly add a flag to make it stricter.

Reply:
Thanks!